### PR TITLE
24 topo speedup

### PR DESCRIPTION
**Pull request by Tklauw-TNO**
Did a thorough review of the changes that were proposed on CI and integrated most of them into this branch. 

Ran the GROW ESDL through the workflow both before and after my changes, which gave very similar results (less than 0.002% difference in MIP gap).

@jimrojerTNO if you have time, please have a look already.

Next up are tests for the proposed changes to heat_mixin.py

### CODE REVIEW

**Comment by jimrojerTNO** [comment]
> Did a thorough review of the changes that were proposed on CI and integrated most of them into this branch.
> 
> Ran the GROW ESDL through the workflow both before and after my changes, which gave very similar results (less than 0.002% difference in MIP gap).
> 
> @jimrojerTNO if you have time, please have a look already.
> 
> Next up are tests for the proposed changes to heat_mixin.py

**Comment by jimrojerTNO** [comment]
@Tklauw-TNO I will have a look. When you say you ran the GROW esdl you mean Rijswijk, right? How long did it now take to run?

**Comment by Tklauw-TNO** [comment]
Yes I mean Rijswijk. The solver took about 30 seconds for both the first and second priority goals each. So the total computation time was in the order of 60-90 seconds (including transcribing the problem).

**Comment by Tklauw-TNO** [comment]
This is indeed with Gurobi, I'll run the Highs solver over lunch, will update you this afternoon.

The current options are:
'neglect_pipe_heat_losses' = False
'heat_loss_disconnected_pipe' = True

File: src/rtctools_heat_network/heat_mixin.py [hunk 1]
```diff
@@ -1942,18 +2048,26 @@ def __flow_direction_path_constraints(self, ensemble_member):
...
             )
 
             # Flow direction is the same for In and Out. Note that this
             # ensures that the heat going in and/or out of a pipe is more than
             # its heat losses.
-            constraints.append(((heat_out - big_m * flow_dir) / constraint_nominal, -np.inf, 0.0))
+            constraints.append(((heat_out - big_m * (flow_dir + is_disconnected)) / constraint_nominal, -np.inf, 0.0))
             constraints.append(
-                ((heat_out + big_m * (1 - flow_dir)) / constraint_nominal, 0.0, np.inf)
+                ((heat_out + big_m * (1 - flow_dir + is_disconnected)) / constraint_nominal, 0.0, np.inf)
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/heat_mixin.py:2070]: Potentially this constraint gives problems. As flow_dir is already free when is_disconnected==1

**Comment by Tklauw-TNO** [comment]
@jimrojerTNO I've completed a first test for the variables. This test now checks the topo variables (pipe sizes, costs, heat losses, etc.) and the ordering variables for the example PipeDiameterSizing. I don't think all other variables defined for the optimizer in HeatMixin are covered by a test yet, but I figured that's outside of the scope of this Pull Request.

File: src/rtctools_heat_network/workflows/grow_workflow.py [hunk 2]
```diff
@@ -518,7 +521,7 @@ def solver_options(self):
 def main(runinfo_path, log_level):
     logger.info("Run Scenario Sizing")
     _ = run_optimization_problem(
-        EndScenarioSizingHIGHS,
+        EndScenarioSizing,
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/workflows/grow_workflow.py:524]: let's put the default here back to highs

File: tests/test_topo_constraints.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                        err_msg=f"Expected the heat loss order var for {p} and {pc=} to be 1.0, "
+                        f"since {chosen_pc=} with lower heat losses",
+                    )
+
+        for pc_name, total_count in pc_sums.items():
+            count_var_value = self.results[f"{pc_name}__global_pipe_class_count"]
+            np.testing.assert_almost_equal(
+                count_var_value,
+                total_count,
+                6,
```
Comment (jimrojerTNO) [inline tests/test_topo_constraints.py:183]: I don't get the 6 here, this now is tested if accurate up to the 6th decimal. Why not just use the default settings?

File: src/rtctools_heat_network/heat_mixin.py [hunk 4]
```diff
@@ -200,20 +221,20 @@ def _get_min_bound(bound):
 
         # Mixed-integer formulation applies only to hot pipes, not to cold
         # pipes.
-        for p in self.hot_pipes:
-            flow_dir_var = f"{p}__flow_direct_var"
+        for pc in self.hot_pipes:
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/heat_mixin.py:224]: why do we change from p to pc here in the naming?

File: src/rtctools_heat_network/heat_mixin.py [hunk 5]
```diff
@@ -3736,8 +4048,8 @@ def __pipe_topology_constraints(self, ensemble_member):
                 p if p in self.hot_pipes else self.cold_to_hot_pipe(p)
             ]
             v = []
-            for pc_var_name in pipe_classes.values():
-                v.append(self.extra_variable(pc_var_name, ensemble_member))
+            for pc in pipe_classes.values():
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/heat_mixin.py:4051]: In this part of the code we set the heat_loss variable, I think for clarity it is better if we put these in a separate function, as they need to be set independently of topology optimization.

**Review by jimrojerTNO — COMMENTED** [review]
@Tklauw-TNO Looks really good. Very happy with the tests, they now actually cover the full set of variables set. 

Tomorrow I will make a commit with my suggested changes in it, I think after that we can merge.

**Review by jimrojerTNO — APPROVED** [review]

File: src/rtctools_heat_network/heat_mixin.py (hunk 5 above)
Comment (Tklauw-TNO) [reply src/rtctools_heat_network/heat_mixin.py:4056]: Yes, good suggestion :)

File: src/rtctools_heat_network/heat_mixin.py (hunk 4 above)
Comment (Tklauw-TNO) [reply src/rtctools_heat_network/heat_mixin.py]: That was just a mistake, good find :)

File: src/rtctools_heat_network/workflows/grow_workflow.py (hunk 2 above)
Comment (Tklauw-TNO) [reply src/rtctools_heat_network/workflows/grow_workflow.py]: Yes, good idea.

File: tests/test_topo_constraints.py (hunk 3 above)
Comment (Tklauw-TNO) [reply tests/test_topo_constraints.py]: I don't think there's a reason. I probably copied the 6 over from the test_pipe_class_var. I'll see if I the default suffices there too.

**Comment by Tklauw-TNO** [comment]
> @Tklauw-TNO Looks really good. Very happy with the tests, they now actually cover the full set of variables set.
> 
> I pushed some changes for the comments that I made, shall we merge?

Yes, I had some more minor changes based on your feedback. If these don't break the pipelines, I'll merge :).

Thanx for the quick review btw, hope doing those on a Sunday doesn't become the standard for you ;).

**Pull request merged**